Add basic namespace resources and data object

ashulmanWeWork opened the pull request.

sshah-wework commented on src/main/java/marquez/api/Namespace.java:
```diff
@@ -0,0 +1,75 @@
+package marquez.api;
```
I think we are moving non-request/response entities to `core/models`?

ashulmanWeWork replied:
I think so, but I'd rather make that refactoring change when we move all the entities over (e.g. Owner, JobRun, etc.).

sshah-wework commented on src/main/java/marquez/resources/NamespaceResource.java:
```diff
@@ -0,0 +1,82 @@
+package marquez.resources;
```
You can bring up the test coverage with tests for this resource.

ashulmanWeWork commented on src/test/java/marquez/api/JobRunBaseTest.java:
```diff
@@ -40,7 +40,7 @@ public static void setUp() {
               handle.execute(
                   format(
-                      "insert into jobs (guid, name, category, description) values "
+                      "insert into jobs (guid, name, category, DESCRIPTION) values "
```
good catch, yep!

sshah-wework commented on src/test/java/marquez/api/NamespaceIntegrationTest.java:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Test
+  public void testValidNamespace() throws NamespaceException {
```
this should be in a `NamespaceResource` test suite

sshah-wework commented on src/test/java/marquez/api/NamespaceIntegrationTest.java:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Test
+  public void testGetNamespaceRequestErrorHandling() throws Exception {
```
`NamespaceResource` test

sshah-wework commented on src/test/java/marquez/api/NamespaceIntegrationTest.java:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Test
+  public void testGetAllNamespaceWithSingleResultSet() throws NamespaceException {
```
`NamespaceResource` test

sshah-wework commented on src/test/java/marquez/api/NamespaceIntegrationTest.java:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Test
+  public void testGetAllNamespaceWithMultipleResultSet() throws NamespaceException {
```
`NamespaceResource` test

sshah-wework commented on src/test/java/marquez/api/NamespaceIntegrationTest.java:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Test
+  public void testGetAllNamespacesErrorHandling() throws NamespaceException {
```
`NamespaceResource` test

sshah-wework commented on src/test/java/marquez/api/NamespaceIntegrationTest.java:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  @Test
+  public void testGetAllNamespacesWithoutTrailingSlash() {
```
is this test needed? we won't ever change how the trailing slash is treated, it's really just testing how dropwizard works

ashulmanWeWork replied:
Good point. Will remove.

ashulmanWeWork commented on src/main/java/marquez/resources/NamespaceResource.java:
```diff
@@ -0,0 +1,108 @@
+      }
+    } catch (Exception e) {
+      return Response.status(Response.Status.BAD_REQUEST)
+          .entity(new ErrorResponse("Not able to create namespace: " + e.getLocalizedMessage()))
```
Agree that the stacktrace might be too much, but wouldn't it be helpful to have more detail as opposed to a generic message?

sshah-wework replied:
We can and should log the exception stacktrace. But for the caller, this is way too much info. Something unexpected went wrong on our side, there's nothing they should have to know / do on their side.

ashulmanWeWork replied:
Resolved.

ashulmanWeWork commented on src/main/java/marquez/resources/NamespaceResource.java:
```diff
@@ -0,0 +1,108 @@
+  @Path("/{namespace}")
+  public Response create(
+      @PathParam("namespace") String namespace, @Valid CreateNamespaceRequest request) {
+    LOG.info("Trying to create about namespace " + namespace);
```
Removed these

ashulmanWeWork commented on src/main/java/marquez/resources/NamespaceResource.java:
```diff
@@ -0,0 +1,108 @@
+      return Response.status(Response.Status.OK)
+          .entity(
+              new CreateNamespaceResponse(
+                  n.getOwnerName(), request.getDescription(), n.getCreatedAt()))
```
Noted - will keep that in mind.

ashulmanWeWork commented:
I noticed that the response from the getAllNamespaces was a list, with each item being the toString representation of a Namespace. Not only is that not elegant and not particularly usable, it leaks GUID info that we wish to keep internal unless absolutely necessary.

In my latest patch, I added a mapper from a Namespace to a GetNamespaceResponse. I also adjusted my tests. This required also adding an equality operator to GetNamespaceResponse.

sshah-wework commented:
happy to review again whenever the changes are in -- thanks for the quick turn around! This is coming along nicely 🎉

ashulmanWeWork commented:
@sshah-wework, I updated the logging on errors. Should be ready for review :).

wslulciuc commented on src/main/java/marquez/api/GetAllNamespacesResponse.java:
```diff
@@ -0,0 +1,25 @@
+
+public class GetAllNamespacesResponse {
+
+  List<GetNamespaceResponse> namespaces = new ArrayList<>();
```
Wouldn't `GetAllNamespacesResponse` contain a list of namespaces, not a list of `GetNamespaceResponse`. You can view `XResponse` or `XRequest` as strictly the HTTP body.

wslulciuc replied:
Oh, I see why you used `GetNamespaceResponse`, there are fields we don't want to return. We still wouldn't use `GetNamespaceResponse` given it's a bit awkward, wondering if the naming convention `XResponse` / `XRequest` adds much benefits.

ashulmanWeWork replied:
Ok so removing the HTTP verb from the name? That might work in the case that the response for a PUT and a GET would be the same. However, looking at the api spec, that's not true in some cases (e.g. jobRun PUT vs. jobRun GET).

ashulmanWeWork replied:
I'm totally on-board that the naming/semantics are a little funky. Open to suggestions.

ashulmanWeWork replied:
Fixed in latest version.

wslulciuc commented on src/main/java/marquez/api/GetAllNamespacesResponse.java:
```diff
@@ -0,0 +1,25 @@
+
+  public GetAllNamespacesResponse(@JsonProperty("namespaces") List<Namespace> namespaceList) {
+    for (Namespace n : namespaceList) {
+      // TODO: Adjust this once we have a mapper in place
```
You wouldn't need a mapper and can avoid transformations by:
```java
this.namespaces = namespaces
```

ashulmanWeWork replied:
They're different types right? Are you suggesting a more elegant way transform from one to another, or are you suggesting changing the input type?

wslulciuc commented on src/main/java/marquez/api/GetAllNamespacesResponse.java:
```diff
@@ -0,0 +1,25 @@
+
+  List<GetNamespaceResponse> namespaces = new ArrayList<>();
+
+  public GetAllNamespacesResponse(@JsonProperty("namespaces") List<Namespace> namespaceList) {
```
Very minor, but I think `ListNamespacesResponse` instead of `GetAllNamespacesResponse` translate more to the action of the caller.

ashulmanWeWork replied:
Sounds good. Less awkward-sounding this way.

wslulciuc commented on src/main/java/marquez/api/GetNamespaceResponse.java:
```diff
@@ -0,0 +1,53 @@
+
+  @JsonCreator
+  public GetNamespaceResponse(
+      final String name,
```
This works without `@JsonProperty("name")`?

ashulmanWeWork replied:
Good point. I'll add it.

ashulmanWeWork replied:
We'll get more clarity on this once we have integration tests that work with data.

wslulciuc commented on src/test/java/marquez/NamespaceBaseTest.java:
```diff
@@ -0,0 +1,33 @@
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+
+public class NamespaceBaseTest {
```
Seems like just boilerplate stuff you're abstracting away? Unless it's commonly shared across tests, I'd just have this defined in `NamespaceIntegrationTest`.

wslulciuc replied:
So there is some reuse in `NamespaceResourceTest`, but that also means `AppWithPostgresRule` is defined without ever being used, which does add to the overall run of the tests for `NamespaceResource` given we are setting up postgres.

ashulmanWeWork replied:
Good catch. Moved the app rule down to the class that needs it. Most everything else is used in both.
